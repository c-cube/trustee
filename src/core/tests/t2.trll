
; basic factorial with a loop

(def "fact" {
  (rstore 0)
  1 (rstore 1) ; sum = 1
  :loop
    (rload 0) 1 leq ; n <= 1
  (jif :ret)
    (rload 0) (rload 1) mult (rstore 1) ; sum *= n
    (rload 0) sub1 (rstore 0) ; n--
    (jmp :loop)
  :ret (rload 1) ret
})

(eval 5 (tforce "fact") call show)
