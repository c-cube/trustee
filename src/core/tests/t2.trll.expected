(def fact =
 result: chunk[2 regs] {
           :instrs
            0        (rstore 0)
            1        (int 1)
            2        (rstore 1)
            3        (rload 0)
            4        (int 1)
            5        leq
            6        (jif 15)
            7        (rload 0)
            8        (rload 1)
            9        mult
            10       (rstore 1)
            11       (rload 0)
            12       sub1
            13       (rstore 0)
            14       (jmp 3)
            15       (rload 1)
            16       ret
           :locals
           
         })
120
(eval: )
(eval: 
   3628800)
